File: Medium/86_Partition_List/Solution.py
from typing import List, Optional
from devlibs.singly_linkedlist import ListNode, printSinglyLinkList


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def partition(self, head: Optional[ListNode], x: int) -> Optional[ListNode]:
        smaller = ListNode(0)
        larger = ListNode(0)

        smaller_head = smaller
        larger_head = larger

        while head:
            if head.val < x:
                smaller.next = head
                smaller = smaller.next
            else:
                larger.next = head
                larger = larger.next

            head = head.next

        larger.next = None
        smaller.next = larger_head.next

        return smaller_head.next


if __name__ == "__main__":
    solution = Solution()

    # test cases 1
    head = ListNode(1, ListNode(4, ListNode(3, ListNode(2, ListNode(5, ListNode(2))))))
    x = 3
    result1 = solution.partition(head, x)
    printSinglyLinkList(result1)  # Expected output: [1, 2, 2, 4, 3, 5]

    # test cases 2
    head2 = ListNode(2, ListNode(1))
    x2 = 2
    result2 = solution.partition(head2, x2)
    printSinglyLinkList(result2)  # Expected output: [1, 2]

File: Medium/86_Partition_List/Solution.java


public class Solution {
    public ListNode partition(ListNode head, int x) {
        ListNode smaller = new ListNode(0);
        ListNode larger = new ListNode(0);

        ListNode smallerHead = smaller;
        ListNode largerHead = larger;

        while (head != null) {
            if (head.val < x) {
                smaller.next = head;
                smaller = smaller.next;
            } else {
                larger.next = head;
                larger = larger.next;
            }
            head = head.next;
        }

        larger.next = null;
        smaller.next = largerHead.next;

        return smallerHead.next;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();

        // test cases 1
        ListNode head1 = new ListNode(1, new ListNode(4, new ListNode(3, new ListNode(2, new ListNode(5, new ListNode(2))))));
        ListNode result1 = solution.partition(head1, 3);
        ListNode current1 = result1;
        while (current1 != null) {
            System.out.print(current1.val + " ");
            current1 = current1.next;
        }
        System.out.println();

        // test cases 2
        ListNode head2 = new ListNode(2, new ListNode(1));
        ListNode result2 = solution.partition(head2, 2);
        ListNode current2 = result2;
        while (current2 != null) {
            System.out.print(current2.val + " ");
            current2 = current2.next;
        }
        System.out.println();
        
    }
}

// Definition for singly-linked list.
class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}